Update the contribution guide (#1868)
beeware/toga: docs/how-to/contribute-code.rst

<https://github.com/beeware/toga/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22>`__
. These have been identified as problems that have a known cause, and we believe
the fix *should* be relatively simple (although we might be wrong in our
analysis).

We don't have any formal process of "claiming" or "assigning" issues; if you're
interested in a ticket, leave a comment that says you're working on it. If
there's an existing comment that says someone is working on the issue, and that
comment is recent, then leave a comment asking if they're still working on the
issue. If you don't get a response in a day or two, you can assume the issue is
available. If the most recent comment is more than a few weeks old, it's
probably safe to assume that the issue is still available to be worked on.

If an issue is particularly old (more than 6 months), it's entirely possible
that the issue has been resolved, so the first step is to verify that you can
reproduce the problem. Use the information provided in the bug report to try and
reproduce the problem. If you can't reproduce the problem, report what you have
found as a comment on the ticket, and pick another ticket.

> If a bug report has no comments from anyone other that the original reporter,
> the issue needs need to be triaged. Triaging a bug involves taking the

Reviewer: ```suggestion
If a bug report has no comments from anyone other than the original reporter,
the issue needs to be triaged. Triaging a bug involves taking the
```

experience writing iOS apps), or you'd *like* to have that experience, you might
want to look into writing tests for a platform backend. We want to get to 100%
coverage for all the backend APIs, but we're a long way from that goal.

The platform backends are tested using a testbed app. Details on how to run the
testbed app for a given platform :ref:`can be found below <run-testbed>`.

Contribute improvements to documentation
----------------------------------------

We've got a :doc:`separate contribution guide <./contribute-docs>` for
documentation contributions. This covers how to set up your development
environment to build Toga's documentation, and separate ideas for what to work
on.

Implement a platform native widget
----------------------------------

If the core library already specifies an interface for a widget, but the widget
isn't implemented on your platform of choice, implement that interface. The
:doc:`supported widgets by platform </reference/widgets_by_platform>` table can
show you the widgets that are missing on various platforms. You can also look
for log messages in a running app (or the direct ``factory.not_implemented()``

> function calls that produce those log messages). At present, the web backend is

Reviewer: ```suggestion
function calls that produce those log messages). At present, the web backend
```

      (venv) C:\...>tox -e py-core

> You should get some output indicating that tests have been run. You may seek

Reviewer: ```suggestion
You should get some output indicating that tests have been run. You may see
```

That is, one more test has been executed, resulting in one less missing line
in the coverage results.

When you're developing your new test, it may be helpful to run *just* that one
test. To do this, you can pass in the name of a specific test file (or a
specific test, using `pytest specifiers
<https://docs.pytest.org/en/latest/how-to/usage.html>`__):

.. tabs::

  .. group-tab:: macOS

    .. code-block:: bash

      (venv) $ tox -e py-core -- tests/path_to_test_file/test_some_test.py

  .. group-tab:: Linux

    .. code-block:: bash

      (venv) $ tox -e py-core -- tests/path_to_test_file/test_some_test.py

  .. group-tab:: Windows

    .. code-block:: doscon

      (venv) C:\...>tox -e py-core -- tests/path_to_test_file/test_some_test.py

> These tests paths are relative to the ``core`` directory. You'll still get a

Reviewer: ```suggestion
These test paths are relative to the ``core`` directory. You'll still get a
```

been laid out correctly inside the Toga window, is drawn using the right font,
and has any other other appropriate physical properties or internal state.

The probe also provides a programmatic interface for interacting *with* a
widget. For example, in order to test a button, you need to be able to press
that button; the probe API provides an API to simulate that press. This allows
the testbed to verify that the correct callbacks will be invoked when a button
is pressed. These interactions are performed by generating events in the GUI
framework being tested.

The widget probe also provides a ``redraw()`` method. GUI libraries don't always
immediately apply changes visually, as graphical changes will often be batched
so that they can be applied in a single redraw. To ensure that any visual
changes have been applied before a test asserts the properties of the app, a
test case can call ``await probe.redraw()``. This guarantees that any
outstanding redraw events have been processed. These ``redraw()`` requests are
also used to implement slow mode - each redraw is turned into a 1 second sleep.

If a widget doesn't have a probe for a given widget, the testbed will

> automatically skip the tests for that widget. If a widget hasn't implemented a

Reviewer: This isn't true anymore: https://github.com/beeware/toga/pull/1830#issuecomment-1485076558.

Author: Oh yeah - I forgot we did that. Updated.

request number. You can then push an update to the pull request, adding the
change note with the assigned number.

It's not just about coverage!
-----------------------------

> Although we're always trying to improve test coverage is always the goal, the

Reviewer: ```suggestion
Although we're always trying to improve test coverage, the
```
